### ISSUE TITLE
Can we make enum with ADT?

### ISSUE THREAD

**Issue by danishin** [tminglei/slick-pg#257]
I'm wondering if Algebraic Data Type pattern can be used to create `enum`, by ADT I mean `sealed trait` with a bunch of `case object/class`.

I find myself very often duplicating code in `object SomeEnumType extends Enumeration` that I create for the sole purpose of passing to `createEnumJdbcType` and a set of ADT (`sealed trait SomeType`) that I need to manually map to each `Value` in above `SomeEnumType`. I use ADT throughout my code since `Enumeration` is simply too limited - not a product type. 

Although since I'm not familiar with how the mapping for `enum` works in slick-pg, maybe this is outright not possible. Can we make `enum` with ADT?

**Comment by tminglei** [comment]
@danishin how can I get all the sub types of a `sealed` trait in scala code?

**Comment by danishin** [comment]
How about something like this?

```
  trait PgEnum[E] {
    type F <: PgEnum[E] { type F = E }
    def allValues: List[F]
  }

  sealed trait Gender extends PgEnum[Gender] {
    import Gender._

    type F = Gender
    def allValues = List(Male, Female)
  }

  object Gender {
    case object Male extends Gender
    case object Female extends Gender
  }
```

I know this requires a little bit of boilerplates from the user's end but ADT using `case object` is worth the grunt I think. And mixing `case object` is prevented through F-bounded polymorphism there.

**Comment by tminglei** [comment]
It's not convenient. I don't think it's a good way.

**Comment by mdedetrich** [comment]
Would probably make more sense to add support to a project like https://github.com/lloydmeta/enumeratum (either on their side or on the slick-pg side)

**Comment by tminglei** [comment]
@mdedetrich ok, I'll think about it.

**Comment by tminglei** [comment]
@danishin pls check #286, is this your expected?

**Comment by danishin** [comment]
@tminglei Yes! That's exactly what I expected although with drastically less duplication. Could you publish with that PR?

**Comment by danishin** [comment]
Feel free to close it either before or after you publish!

**Comment by tminglei** [comment]
@danishin released slick-pg v0.14.2, pls check and use it. _Thanks!_
